File: backend/routers/status.py
import asyncio

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from downloader import dm

router = APIRouter()


@router.get("/api/status")
async def api_status():
    return await dm.global_stats()


@router.websocket("/ws/progress")
async def ws_progress(ws: WebSocket):
    await ws.accept()
    try:
        while True:
            await ws.send_json({
                "type": "progress",
                "data": dm.all_progress(),
                "aria2_ok": dm.is_running,
            })
            await asyncio.sleep(1)
    except WebSocketDisconnect:
        pass
    except Exception:
        pass

File: src/routes/status.ts
import type { FastifyInstance } from 'fastify';
import { dm } from '../downloader.js';

export function registerStatusRoutes(app: FastifyInstance): void {
  app.get('/api/status', async () => dm.globalStats());

  app.get('/ws/progress', { websocket: true }, (socket) => {
    const timer = setInterval(() => {
      try {
        socket.send(JSON.stringify({
          type: 'progress',
          data: dm.allProgress(),
          aria2_ok: dm.isRunning,
        }));
      } catch {
        clearInterval(timer);
      }
    }, 1000);
    socket.on('close', () => clearInterval(timer));
    socket.on('error', () => clearInterval(timer));
  });
}
